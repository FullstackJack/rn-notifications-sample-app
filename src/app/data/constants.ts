import {Theme} from '@react-navigation/native';

export const CustomTheme: Theme = {
  dark: true,
  colors: {
    primary: 'rgb(10, 132, 255)',
    background: 'rgb(13, 20, 34)',
    card: 'rgb(22, 31, 43)',
    text: 'rgb(229, 229, 231)',
    border: 'rgb(39, 39, 41)',
    notification: 'rgb(255, 69, 58)',
  },
};
