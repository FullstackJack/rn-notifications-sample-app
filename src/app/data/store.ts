import {
  PreloadedState,
  combineReducers,
  configureStore,
} from '@reduxjs/toolkit';
import notifications from '../../modules/notifications/data/notificationsSlice';

// Create the root reducer independently to obtain the RootState type
const rootReducer = combineReducers({
  notifications,
});

export const store = configureStore({
  reducer: rootReducer,
});

export function setupStore(preloadedState?: PreloadedState<RootState>) {
  return configureStore({
    reducer: {notifications},
    preloadedState,
  });
}

// Infer the `RootState` and `AppDispatch` types from the store itself
export type RootState = ReturnType<typeof rootReducer>;
export type AppStore = ReturnType<typeof setupStore>;
// Inferred type: {notifications: NotificationsState, ...}
export type AppDispatch = typeof store.dispatch;
