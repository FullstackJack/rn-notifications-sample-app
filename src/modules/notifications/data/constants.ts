export const ACTION_FRIEND_REQUEST = 'friend_request';
export const ACTION_FRIEND_REQUEST_DECLINED = 'friend_request_declined';
export const ACTION_FRIEND_REQUEST_ACCEPTED = 'friend_request_accepted';
export const ACTION_MENTION = 'mention';
export const ACTION_COMMUNITY_INVITE = 'community_invite';
export const NOTIFICAION_INTERVAL = 30000;
