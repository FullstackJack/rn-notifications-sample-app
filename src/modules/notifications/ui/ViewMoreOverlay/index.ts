export {default as ViewMoreOverlay} from './ViewMoreOverlay';
