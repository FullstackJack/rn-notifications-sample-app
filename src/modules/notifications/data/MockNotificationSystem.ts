import {useEffect, useRef} from 'react';
import {useDispatch} from 'react-redux';
import {addNotification} from './notificationsSlice';
import {faker} from '@faker-js/faker';
import {v4 as uuidv4} from 'uuid';
import {
  ACTION_COMMUNITY_INVITE,
  ACTION_FRIEND_REQUEST,
  ACTION_FRIEND_REQUEST_ACCEPTED,
  ACTION_FRIEND_REQUEST_DECLINED,
  ACTION_MENTION,
  NOTIFICAION_INTERVAL,
} from './constants';

export default function MockNotificationSystem() {
  const dispatch = useDispatch();
  const intervalRef = useRef<NodeJS.Timeout | undefined>();

  useEffect(() => {
    intervalRef.current = setInterval(() => {
      dispatch(
        addNotification({
          notification: {
            id: uuidv4(),
            user: {
              name: faker.person.firstName(),
              id: uuidv4(),
            },
            createdAt: Date.now() - new Date().getTimezoneOffset() * 60000,
            isRead: false,
            action: faker.helpers.arrayElement([
              ACTION_COMMUNITY_INVITE,
              ACTION_MENTION,
              ACTION_FRIEND_REQUEST,
              ACTION_FRIEND_REQUEST_ACCEPTED,
              ACTION_FRIEND_REQUEST_DECLINED,
            ]),
            community: {
              id: '0d568bc4-057c-4bea-9746-dcc64e6f1a7d',
              name: 'Root',
            },
            channel: faker.helpers.arrayElement([
              {
                id: '0d568bc4-057c-4bea-9746-dcc64e6f1a7d',
                name: 'Infrustruture',
              },
              {
                id: '4b17ca83-4e1d-4efe-ab81-85ae1e3526cc',
                name: 'General',
              },
              {
                id: '763a9f14-30b4-4435-b4d2-20f4e824f963',
                name: 'Client',
              },
              {
                id: '6f6a7df0-2c61-4663-ac3c-73cda5ca6d2a',
                name: 'Pinball',
              },
            ]),
          },
        }),
      );
    }, NOTIFICAION_INTERVAL);
    return function cleanup() {
      if (intervalRef.current) {
        clearInterval(intervalRef.current);
      }
    };
  }, [dispatch]);
  return null;
}
