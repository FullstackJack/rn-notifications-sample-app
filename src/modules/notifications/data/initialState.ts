import {NotificationsState} from './notificationsSlice';
import {
  ACTION_FRIEND_REQUEST,
  ACTION_FRIEND_REQUEST_ACCEPTED,
  ACTION_FRIEND_REQUEST_DECLINED,
  ACTION_MENTION,
} from './constants';

export const initialState: NotificationsState = {
  newNotifications: {},
  newNotificationsCount: 0,
  notifications: {
    n0: {
      id: 'n0',
      action: ACTION_MENTION,
      createdAt: 1699873010725,
      channel: {
        name: 'Client',
        id: 'ch123',
      },
      community: {
        name: 'Root',
        id: 'c123',
      },
      isRead: false,
      user: {
        name: 'Vinny',
        id: 'u123',
      },
    },
    n1: {
      id: 'n1',
      action: ACTION_MENTION,
      createdAt: 1699873227983,
      channel: {
        name: 'General',
        id: 'ch123',
      },
      community: {
        name: 'Root',
        id: 'c123',
      },
      isRead: false,
      user: {
        name: 'Jesse',
        id: 'u123',
      },
    },
    n2: {
      id: 'n2',
      action: ACTION_MENTION,
      createdAt: 1699860432773,
      channel: {
        name: 'Infrastructure',
        id: 'ch123',
      },
      community: {
        name: 'Root',
        id: 'c123',
      },
      isRead: false,
      user: {
        name: 'Vinny',
        id: 'u123',
      },
    },
    n3: {
      id: 'n3',
      action: ACTION_FRIEND_REQUEST_DECLINED,
      createdAt: 1699860432773,
      isRead: false,
      user: {
        name: 'Mark',
        id: 'u123',
      },
    },
    n4: {
      id: 'n4',
      action: ACTION_FRIEND_REQUEST_ACCEPTED,
      createdAt: 1699860432773,
      isRead: false,
      user: {
        name: 'Jesse',
        id: 'u123',
      },
    },
    n5: {
      id: 'n5',
      action: ACTION_FRIEND_REQUEST,
      createdAt: 1699801790151,
      isRead: false,
      user: {
        name: 'Vinny',
        id: 'u123',
      },
    },
    n6: {
      id: 'n6',
      action: ACTION_FRIEND_REQUEST,
      createdAt: 1699951353176,
      isRead: false,
      user: {
        name: 'Matt',
        id: 'u123',
      },
    },
  },
};
