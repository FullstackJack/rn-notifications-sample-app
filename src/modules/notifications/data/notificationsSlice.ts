import {PayloadAction, createSlice} from '@reduxjs/toolkit';
import {INotification} from '../types';
import {initialState} from './initialState';

export interface NotificationsState {
  notifications: Record<string, INotification>;
  newNotifications: Record<string, INotification>;
  newNotificationsCount: number;
}

export const notificationsSlice = createSlice({
  name: 'notifications',
  initialState,
  reducers: {
    // NOTE: Redux Toolkit allows us to write "mutating" logic in reducers. It
    // doesn't actually mutate the state because it uses the Immer library,
    // which detects changes to a "draft state" and produces a brand new
    // immutable state based off those changes
    markAsRead: (state, action: PayloadAction<{id: string}>) => {
      state.notifications[action.payload.id].isRead = true;
    },
    addNotification: (
      state,
      action: PayloadAction<{notification: INotification}>,
    ) => {
      state.newNotifications[action.payload.notification.id] =
        action.payload.notification;
      state.newNotificationsCount += 1;
    },
    showNewNotifications: state => {
      state.notifications = {...state.notifications, ...state.newNotifications};
      state.newNotifications = {};
      state.newNotificationsCount = 0;
    },
  },
});

// NOTE: Action creators are generated for each case reducer function
export const {markAsRead, addNotification, showNewNotifications} =
  notificationsSlice.actions;

export default notificationsSlice.reducer;
