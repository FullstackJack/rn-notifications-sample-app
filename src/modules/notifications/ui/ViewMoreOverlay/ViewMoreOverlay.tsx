import React from 'react';

import {useDispatch, useSelector} from 'react-redux';
import {RootState} from '../../../../app/data/store';
import {StyleSheet, View} from 'react-native';
import {FlatButton} from '../../../../common/ui/FlatButton';
import {showNewNotifications} from '../../data/notificationsSlice';

interface IViewMoreOverlay {
  onViewMorePress?: () => void;
}

export default function ViewMoreButton({onViewMorePress}: IViewMoreOverlay) {
  const dispatch = useDispatch();
  const newNotificationsCount = useSelector(
    (state: RootState) => state.notifications.newNotificationsCount,
  );

  return newNotificationsCount ? (
    <View style={styles.container}>
      <FlatButton
        text={`${newNotificationsCount} New Notifications`}
        style={styles.button}
        textStyle={styles.buttonText}
        onPress={() => {
          dispatch(showNewNotifications());
          onViewMorePress?.();
        }}
      />
    </View>
  ) : null;
}

const styles = StyleSheet.create({
  container: {
    position: 'absolute',
    height: 50,
    backgroundColor: 'rgba(0, 0, 0, 0.5)',
    bottom: 0,
    left: 0,
    right: 0,
    alignItems: 'center',
    justifyContent: 'center',
  },
  button: {alignSelf: 'center', backgroundColor: '#993333'},
  buttonText: {color: 'white'},
});
