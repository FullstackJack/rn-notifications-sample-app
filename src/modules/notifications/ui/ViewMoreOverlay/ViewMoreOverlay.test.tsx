import React from 'react';
import {expect, test} from '@jest/globals';
import {screen} from '@testing-library/react-native';
import {renderWithProviders} from '../../../../common/utils/test-utils';
import ViewMoreOverlay from './ViewMoreOverlay';
import {NotificationsState} from '../../data/notificationsSlice';

test('renders correctly', async () => {
  renderWithProviders(<ViewMoreOverlay />, {
    preloadedState: {
      notifications: {newNotificationsCount: 10} as NotificationsState,
    },
  });

  expect(await screen.findByText('10 New Notifications')).toHaveTextContent(
    '10 New Notifications',
  );

  expect(screen.toJSON()).toMatchSnapshot();
});
