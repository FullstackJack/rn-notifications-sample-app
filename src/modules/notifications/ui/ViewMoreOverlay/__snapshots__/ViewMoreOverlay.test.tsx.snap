// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "rgba(0, 0, 0, 0.5)",
      "bottom": 0,
      "height": 50,
      "justifyContent": "center",
      "left": 0,
      "position": "absolute",
      "right": 0,
    }
  }
>
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "alignItems": "center",
          "alignSelf": "flex-start",
          "backgroundColor": "white",
          "borderColor": "transparent",
          "borderRadius": 50,
          "borderStyle": "solid",
          "borderWidth": 1,
          "justifyContent": "center",
          "paddingHorizontal": 16,
          "paddingVertical": 8,
          "width": "auto",
        },
        {
          "alignSelf": "center",
          "backgroundColor": "#993333",
        },
        {
          "backgroundColor": "#993333",
        },
      ]
    }
  >
    <Text
      style={
        {
          "color": "white",
        }
      }
    >
      10 New Notifications
    </Text>
  </View>
</View>
`;
